proposals/default_optional_arguments/revised_proposal.txt

```diff
@@ -0,0 +1,356 @@
+arguments, if the optional dummy argument is meant to be used in
+multiple places in the procedure, the programmer is likely to use the
+above pattern, where a "temporary" variable is declared and used in
+place of the dummy argument, disconnecting the implementation from the
+user interface. Furthermore, this requires at least 3 lines of code
+(declaration of c_tmp, initialization of c_tmp, and testing for the
+presence of c) to handle the scenario of a missing optional argument.
+
+The situation for other intents is different. For INTENT(OUT) one
+might like to have a default value, but there is no need for a
+temporary variable to hold it. The example subroutine shows a typical
+usage:
+
+    subroutine do_stuff( ..., status )
+	...
+	integer, intent(out), optional :: status
+	...
+	if (present(status)) status = 0 ! the desired default
+	...
+	if (bad_stuff_happened) then
+	    if (present(status)) then
+	        status = 1 ! Flag that bad stuff happened
+		return
+	    else
+	        error stop 'Bad Stuff Happened in DO_STUFF'
+            end if
+	end if
+	...
+    end subroutine do_stuff
+
+This differs from the INTENT(IN), VALUE, and no intent case: first,
+because one assigns the default value with INTENT(OUT) when the actual
+argument is present, not when it is absent; and second, because the
+assignment of a default involves only one statement rather than the
+three statements for the helper variable of INTENT(IN), VALUE, and no
+intent. For INTENT(INOUT), one typically makes the argument optional
+to avoid the computational expense of calculating the out result, but
+having a default value makes it awkward to determine the presence of
+an optional argument, so the primary purpose of having it be optional
+is defeated.
+
+This proposal addresses the issue for INTENT(IN), VALUE, and no intent
+arguments that checking for the presence of the optional dummy
+argument and using a helper variable is cumbersome and error-prone. We
+will not address default values for INTENT(OUT) OR INTENT(INOUT)
+arguments. The primary benefit of this feature is the  reduction in
+source code needed to handle optional arguments. This benefit is
+greatest in scenarios where the optional argument is used in many
+places in the procedure, and a helper variable is used for its value
+instead. Reduction in needed source code would result in more readable
+and more correct programs.
+
+
+3. Prior art
+
+We are not aware of any Fortran compiler that has implemented default
+arguments. However it is a very common feature of other languages. Of
+interpreted languages widely used in the scientific community:
+Python, IDL, R, and Matlab all have default arguments. Of compiled
+languages C++ and Ada may be the best known with default arguments. Of
+these languages, Ada may be the most pertinent as it has INTENT
+arguments similar to Fortran's. Ada has only defined default arguments
+useful for INTENT(IN), and has not defined it for INTENT(OUT) or
```

Comment:
I don't know Ada. What does it implement for "no intent"?

Reply:
Ada doesn't have that option. For its parameters (the Fortran arguments) you always have to specify INTENT with either IN, OUT, or INOUT.